File: src/odom_transform_node.cpp
#include "robot_localization/odom_transform_node.hpp"
#include "rclcpp/clock.hpp"
#include "rclcpp/logging.hpp"
#include "rclcpp/utilities.hpp"
#include <chrono>

using namespace std::chrono_literals;


OdomTransform::OdomTransform() : Node("odom_transform")
{
    // get parameters
    this->declare_parameter<std::string>("odom_topic");
    this->declare_parameter<std::string>("pose_topic");
    this->declare_parameter<std::string>("target_frame");

    this->get_parameter("odom_topic", odom_topic);
    this->get_parameter("pose_topic", pose_topic);
    this->get_parameter("target_frame", target_frame);

    RCLCPP_INFO(this->get_logger(), "Projecting '%s' to frame '%s' on topic '%s'",
        odom_topic.c_str(), target_frame.c_str(), pose_topic.c_str());

    // initialize tf2
    buffer = std::make_unique<tf2_ros::Buffer>(this->get_clock());
    transform_listener = std::make_shared<tf2_ros::TransformListener>(*buffer);

    // initialize odom subscriber
    odom_sub = this->create_subscription<nav_msgs::msg::Odometry>(odom_topic, 10,
        std::bind(&OdomTransform::odom_callback, this, std::placeholders::_1));

    // initialize pose publisher
    pose_pub = this->create_publisher<geometry_msgs::msg::PoseWithCovarianceStamped>(pose_topic, 10);
}


void OdomTransform::odom_callback(const nav_msgs::msg::Odometry::SharedPtr msg) const
{
    // convert Odometry to Pose
    geometry_msgs::msg::Pose pose_before, pose;
    pose_before.position.x = msg->pose.pose.position.x;
    pose_before.position.y = msg->pose.pose.position.y;
    pose_before.position.z = msg->pose.pose.position.z;
    pose_before.orientation.x = msg->pose.pose.orientation.x;
    pose_before.orientation.y = msg->pose.pose.orientation.y;
    pose_before.orientation.z = msg->pose.pose.orientation.z;
    pose_before.orientation.w = msg->pose.pose.orientation.w;

    pose = buffer->transform(pose_before, target_frame);

    // create PoseWithCovarianceStamped
    geometry_msgs::msg::PoseWithCovariance posec;
    posec.pose = pose;
    posec.covariance = msg->pose.covariance;
    geometry_msgs::msg::PoseWithCovarianceStamped posecs;
    posecs.header = msg->header;
    posecs.header.frame_id = target_frame;
    posecs.pose = posec;

    // publish pose
    pose_pub->publish(posecs);
}


int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<OdomTransform>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: scripts/odom_transform_node.py
#! /usr/bin/env python3

import numpy as np
from dh_transform import transform

# ros
import rclpy
from rclpy.node import Node
from tf2_ros import Buffer, TransformException, TransformListener

# msgs
from nav_msgs.msg import Odometry
from geometry_msgs.msg import PoseWithCovarianceStamped, TransformStamped




class OdomTransform(Node):
    def __init__(self):
        super().__init__('odom_transform')
        self.counter = 0
        # get parameters
        self.declare_parameter('odom_topic',   '/t265_camera/odom')
        self.declare_parameter('pose_topic',   '/t265_camera/odom/to_base')
        self.declare_parameter('target_frame', 'base_link')

        self.odom_topic = self.get_parameter('odom_topic').value
        self.pose_topic = self.get_parameter('pose_topic').value
        self.target_frame = self.get_parameter('target_frame').value

        self.get_logger().info(f'Projecting {self.odom_topic} to frame {self.target_frame} on topic {self.pose_topic}')

        # initialize tf2
        self.buffer = Buffer()
        self.tf_listener = TransformListener(self.buffer, self)

        # initialize ros communication
        self.odom_sub = self.create_subscription(Odometry, self.odom_topic, self.odom_callback, 1)
        self.pose_pub = self.create_publisher(PoseWithCovarianceStamped, self.pose_topic, 1)


    def odom_callback(self, odom: Odometry) -> type(None):
        if self.counter < 10:
            self.counter += 1
            return
        else:
            self.counter = 0
        # create PoseWithCovarianceStamped from Odometry
        pose                 = PoseWithCovarianceStamped()
        pose.header          = odom.header
        pose.pose            = odom.pose
        pose.pose.covariance = odom.pose.covariance

        # get transform from target_frame to original one
        tf = TransformStamped()
        try:
            tf  = self.buffer.lookup_transform(odom.child_frame_id, self.target_frame, rclpy.time.Time())
        except TransformException:
            self.get_logger().warn(f'Could not get transform from {odom.child_frame_id} to {self.target_frame}')
            return

        pose.pose.pose = transform(pose.pose.pose, tf.transform)
        self.pose_pub.publish(pose)

def main(args=None) -> type(None):
    rclpy.init(args=args)
    odomTransform = OdomTransform()
    rclpy.spin(odomTransform)
    rclpy.shutdown()

if __name__ == '__main__':
    main()
